File: substrate/floating_multi_select_source_attach_quality_twin_compose.py
"""Floating multi-select + source quality → twin chase feed (pure).

live_dispatched / twin_written / remote_fetched / prompts_injected always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.floating_multi_select_source_attach_quality_compose import (
    FloatingMultiSelectSourceAttachQualityCompose,
    FloatingMultiSelectSourceAttachQualityComposeError,
    compose_floating_multi_select_source_attach_quality,
)
from substrate.twin_chase_analysis_feed_compose import (
    TwinChaseAnalysisFeedCompose,
    TwinChaseAnalysisFeedComposeError,
    compose_twin_chase_analysis_feed,
)


class FloatingMultiSelectSourceAttachQualityTwinComposeError(ValueError):
    """Fail-closed validation for multi-select source twin pack."""


@dataclass(frozen=True)
class FloatingMultiSelectSourceAttachQualityTwinCompose:
    session_id: str
    parent_asset_id: str
    multi_source: FloatingMultiSelectSourceAttachQualityCompose
    twin_feed: TwinChaseAnalysisFeedCompose
    pack_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    analysis_written: bool
    remote_fetched: bool
    twin_written: bool
    prompts_injected: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "multi_source": self.multi_source.to_dict(),
            "twin_feed": self.twin_feed.to_dict(),
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "analysis_written": False,
            "remote_fetched": False,
            "twin_written": False,
            "prompts_injected": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": (
                "floating_multi_select_source_attach_quality_twin_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _derive_findings(
    sources: object,
    members: object,
    selected_instance_ids: object,
    cohesive_prompt: object,
) -> list[dict[str, Any]]:
    if not isinstance(sources, list):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "sources must be an array"
        )
    if not isinstance(members, list):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "members must be an array"
        )
    if not isinstance(selected_instance_ids, list):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "selected_instance_ids must be an array"
        )
    selected = {str(x) for x in selected_instance_ids}
    findings: list[dict[str, Any]] = []
    for s in sources:
        if not isinstance(s, dict):
            raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
                "sources items must be objects"
            )
        sid = str(s.get("source_id", "")).strip()
        title = str(s.get("title", "")).strip()
        if not sid or not title:
            raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
                "sources items need source_id and title"
            )
        findings.append(
            {"source_id": f"src-{sid}", "body": title, "kind": "data"}
        )
    for m in members:
        if not isinstance(m, dict):
            continue
        iid = str(m.get("instance_id", "")).strip()
        if iid not in selected:
            continue
        hl = str(m.get("highlight") or "").strip()
        prior = str(m.get("prior_prompt") or "").strip()
        findings_list = m.get("findings")
        first_finding = ""
        if isinstance(findings_list, list) and findings_list:
            first_finding = str(findings_list[0]).strip()
        body = hl or prior or first_finding or iid
        kind = "insight" if first_finding else "question"
        findings.append(
            {"source_id": f"inst-{iid}", "body": body, "kind": kind}
        )
    prompt = str(cohesive_prompt or "").strip()
    if prompt:
        findings.append(
            {
                "source_id": "cohesive-prompt",
                "body": prompt,
                "kind": "question",
            }
        )
    return findings


def compose_floating_multi_select_source_attach_quality_twin(
    *,
    session_id: object,
    parent_asset_id: object,
    members: object,
    selected_instance_ids: object,
    pack_mode: object,
    cohesive_prompt: object,
    operator_ack: object,
    requested_families: object,
    sources: object,
    quality_overall: object,
    would_exceed: object,
    extra_context: object | None = None,
    analysis_kind: object | None = None,
    extra_findings: object | None = None,
    citations: object | None = None,
    derive_citations_from_sources: object | None = None,
    quality_floor: object | None = None,
    operator_override: object | None = None,
    require_both: object | None = None,
    existing_twin_asset_id: object | None = None,
    analysis_excerpt: object | None = None,
    mark_for_prompt_context: object | None = None,
    twin_findings: object | None = None,
    require_both_with_twin: object | None = None,
) -> FloatingMultiSelectSourceAttachQualityTwinCompose:
    """Multi-select+sources + twin feed. Never dispatches/writes twin."""
    if not isinstance(operator_ack, bool):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")

    require_twin = (
        True if require_both_with_twin is None else require_both_with_twin
    )
    if not isinstance(require_twin, bool):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "require_both_with_twin must be boolean when set"
        )

    mark_prompt = (
        True if mark_for_prompt_context is None else mark_for_prompt_context
    )
    if not isinstance(mark_prompt, bool):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "mark_for_prompt_context must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · pack_dispatched=false · merge_executed=false",
        "remote_fetched=false · twin_written=false · prompts_injected=false",
    ]

    try:
        multi_source = compose_floating_multi_select_source_attach_quality(
            session_id=session,
            parent_asset_id=parent,
            members=members,
            selected_instance_ids=selected_instance_ids,
            pack_mode=pack_mode,
            cohesive_prompt=cohesive_prompt,
            operator_ack=operator_ack,
            requested_families=requested_families,
            sources=sources,
            quality_overall=quality_overall,
            would_exceed=would_exceed,
            extra_context=extra_context,
            analysis_kind=analysis_kind,
            extra_findings=extra_findings,
            citations=citations,
            derive_citations_from_sources=derive_citations_from_sources,
            quality_floor=quality_floor,
            operator_override=operator_override,
            require_both=require_both,
        )
    except FloatingMultiSelectSourceAttachQualityComposeError as e:
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            str(e)
        ) from e
    notes.extend(f"[multi_source] {n}" for n in multi_source.notes)

    if twin_findings is not None:
        if not isinstance(twin_findings, list):
            raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
                "twin_findings must be an array when set"
            )
        findings = twin_findings
        notes.append(f"twin_findings={len(findings)} caller-supplied")
    else:
        findings = _derive_findings(
            sources, members, selected_instance_ids, cohesive_prompt
        )
        notes.append(
            f"twin_findings={len(findings)} derived from "
            "sources+selected members+prompt"
        )

    try:
        twin_feed = compose_twin_chase_analysis_feed(
            session_id=session,
            parent_asset_id=parent,
            findings=findings,
            operator_ack=operator_ack,
            analysis_excerpt=analysis_excerpt,
            existing_twin_asset_id=existing_twin_asset_id,
            mark_for_prompt_context=mark_prompt,
        )
    except TwinChaseAnalysisFeedComposeError as e:
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            str(e)
        ) from e
    notes.extend(f"[twin_feed] {n}" for n in twin_feed.notes)

    if require_twin:
        pack_ready = (
            multi_source.pack_ready is True
            and twin_feed.feed_ready is True
            and operator_ack is True
        )
    else:
        pack_ready = operator_ack is True and (
            multi_source.pack_ready is True or twin_feed.feed_ready is True
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — multi-select+sources + twin feed ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — multi-source, twin feed, or operator_ack gate open"
        )

    if (
        multi_source.live_dispatched is not False
        or multi_source.remote_fetched is not False
        or twin_feed.twin_written is not False
        or twin_feed.prompts_injected is not False
    ):
        raise FloatingMultiSelectSourceAttachQualityTwinComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "analysis_written=false",
            "remote_fetched=false",
            "twin_written=false",
            "prompts_injected=false",
            "store_mutated=false",
        )
    )

    return FloatingMultiSelectSourceAttachQualityTwinCompose(
        session_id=session,
        parent_asset_id=parent,
        multi_source=multi_source,
        twin_feed=twin_feed,
        pack_ready=pack_ready,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        analysis_written=False,
        remote_fetched=False,
        twin_written=False,
        prompts_injected=False,
        store_mutated=False,
        notes=tuple(notes),
        authority=(
            "floating_multi_select_source_attach_quality_twin_compose_advisory"
        ),
    )


def format_floating_multi_select_source_attach_quality_twin_summary(
    c: FloatingMultiSelectSourceAttachQualityTwinCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"multi_source_ready={c.multi_source.pack_ready} · "
        f"twin_feed_ready={c.twin_feed.feed_ready} · "
        f"findings={c.twin_feed.finding_count} · "
        f"live_dispatched=false · twin_written=false · remote_fetched=false"
    )


__all__ = [
    "FloatingMultiSelectSourceAttachQualityTwinCompose",
    "FloatingMultiSelectSourceAttachQualityTwinComposeError",
    "compose_floating_multi_select_source_attach_quality_twin",
    "format_floating_multi_select_source_attach_quality_twin_summary",
]

File: apps/reading/src/api/floatingMultiSelectSourceAttachQualityTwinCompose.ts
/**
 * Floating multi-select + source quality → twin chase feed (pure).
 *
 * Operator vision: multi-select cohesive DR with knowledge-dense sources
 * feeds recursive twin note-taker so collective findings and source titles
 * inform future workstation prompts.
 *
 * live_dispatched / twin_written / remote_fetched always false.
 */

import {
  composeFloatingMultiSelectSourceAttachQuality,
  type FloatingMultiSelectSourceAttachQualityCompose,
  type FloatingMultiSelectSourceAttachQualityInput,
} from "./floatingMultiSelectSourceAttachQualityCompose";
import {
  composeTwinChaseAnalysisFeed,
  type ChaseFeedFinding,
  type TwinChaseAnalysisFeedCompose,
} from "./twinChaseAnalysisFeedCompose";

export interface FloatingMultiSelectSourceAttachQualityTwinInput
  extends FloatingMultiSelectSourceAttachQualityInput {
  existing_twin_asset_id?: string | null;
  analysis_excerpt?: string | null;
  mark_for_prompt_context?: boolean;
  twin_findings?: ChaseFeedFinding[] | null;
  require_both_with_twin?: boolean;
}

export interface FloatingMultiSelectSourceAttachQualityTwinCompose {
  session_id: string;
  parent_asset_id: string;
  multi_source: FloatingMultiSelectSourceAttachQualityCompose;
  twin_feed: TwinChaseAnalysisFeedCompose;
  pack_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  analysis_written: false;
  remote_fetched: false;
  twin_written: false;
  prompts_injected: false;
  store_mutated: false;
  notes: string[];
  authority: "floating_multi_select_source_attach_quality_twin_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function deriveFindings(
  input: FloatingMultiSelectSourceAttachQualityTwinInput,
): ChaseFeedFinding[] {
  const findings: ChaseFeedFinding[] = [];
  for (const s of input.sources) {
    findings.push({
      source_id: `src-${s.source_id}`,
      body: s.title,
      kind: "data",
    });
  }
  for (const m of input.members) {
    if (input.selected_instance_ids.includes(m.instance_id)) {
      const body =
        m.highlight?.trim() ||
        m.prior_prompt?.trim() ||
        m.findings?.[0]?.trim() ||
        m.instance_id;
      findings.push({
        source_id: `inst-${m.instance_id}`,
        body,
        kind: m.findings?.length ? "insight" : "question",
      });
    }
  }
  if (input.cohesive_prompt.trim()) {
    findings.push({
      source_id: "cohesive-prompt",
      body: input.cohesive_prompt.trim(),
      kind: "question",
    });
  }
  return findings;
}

export function composeFloatingMultiSelectSourceAttachQualityTwin(
  input: FloatingMultiSelectSourceAttachQualityTwinInput,
): FloatingMultiSelectSourceAttachQualityTwinCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const require_both_with_twin =
    input.require_both_with_twin === undefined
      ? true
      : input.require_both_with_twin;
  if (typeof require_both_with_twin !== "boolean") {
    throw new Error("require_both_with_twin must be boolean when set");
  }

  const mark_for_prompt_context =
    input.mark_for_prompt_context === undefined
      ? true
      : input.mark_for_prompt_context;
  if (typeof mark_for_prompt_context !== "boolean") {
    throw new Error("mark_for_prompt_context must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · pack_dispatched=false · merge_executed=false",
    "remote_fetched=false · twin_written=false · prompts_injected=false",
  ];

  const multi_source = composeFloatingMultiSelectSourceAttachQuality({
    session_id: input.session_id,
    parent_asset_id: input.parent_asset_id,
    members: input.members,
    selected_instance_ids: input.selected_instance_ids,
    pack_mode: input.pack_mode,
    cohesive_prompt: input.cohesive_prompt,
    operator_ack: input.operator_ack,
    extra_context: input.extra_context,
    analysis_kind: input.analysis_kind,
    extra_findings: input.extra_findings,
    requested_families: input.requested_families,
    sources: input.sources,
    citations: input.citations,
    derive_citations_from_sources: input.derive_citations_from_sources,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    require_both: input.require_both,
  });
  notes.push(...multi_source.notes.map((n) => `[multi_source] ${n}`));

  let twin_findings: ChaseFeedFinding[];
  if (input.twin_findings != null) {
    if (!Array.isArray(input.twin_findings)) {
      throw new Error("twin_findings must be an array when set");
    }
    twin_findings = input.twin_findings;
    notes.push(`twin_findings=${twin_findings.length} caller-supplied`);
  } else {
    twin_findings = deriveFindings(input);
    notes.push(
      `twin_findings=${twin_findings.length} derived from sources+selected members+prompt`,
    );
  }

  const twin_feed = composeTwinChaseAnalysisFeed({
    session_id,
    parent_asset_id,
    findings: twin_findings,
    analysis_excerpt: input.analysis_excerpt,
    existing_twin_asset_id: input.existing_twin_asset_id,
    operator_ack: input.operator_ack,
    mark_for_prompt_context,
  });
  notes.push(...twin_feed.notes.map((n) => `[twin_feed] ${n}`));

  let pack_ready = false;
  if (require_both_with_twin) {
    pack_ready =
      multi_source.pack_ready === true &&
      twin_feed.feed_ready === true &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      (multi_source.pack_ready === true || twin_feed.feed_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — multi-select+sources + twin feed ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — multi-source, twin feed, or operator_ack gate open",
    );
  }

  if (
    multi_source.live_dispatched !== false ||
    multi_source.remote_fetched !== false ||
    twin_feed.twin_written !== false ||
    twin_feed.prompts_injected !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("analysis_written=false");
  notes.push("remote_fetched=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("store_mutated=false");

  return {
    session_id,
    parent_asset_id,
    multi_source,
    twin_feed,
    pack_ready,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    analysis_written: false,
    remote_fetched: false,
    twin_written: false,
    prompts_injected: false,
    store_mutated: false,
    notes,
    authority:
      "floating_multi_select_source_attach_quality_twin_compose_advisory",
  };
}

export function formatFloatingMultiSelectSourceAttachQualityTwinSummary(
  c: FloatingMultiSelectSourceAttachQualityTwinCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `multi_source_ready=${c.multi_source.pack_ready} · ` +
    `twin_feed_ready=${c.twin_feed.feed_ready} · ` +
    `findings=${c.twin_feed.finding_count} · ` +
    `live_dispatched=false · twin_written=false · remote_fetched=false`
  );
}
